src/spirc.rs

```diff
@@ -353,3 +392,93 @@ impl SpircInternal {
         format!("hm://remote/user/{}", self.session.username())
     }
 }
+
+struct CommandSender<'a> {
+    spirc_internal: &'a mut SpircInternal,
+    cmd: protocol::spirc::MessageType,
+    recipient: Option<&'a str>,
+    player_state: Option<&'a PlayerState>,
+    state: Option<protocol::spirc::State>,
+}
+
+impl<'a> CommandSender<'a> {
+    fn new(spirc_internal: &'a mut SpircInternal,
+           cmd: protocol::spirc::MessageType)
+           -> CommandSender {
+        CommandSender {
+            spirc_internal: spirc_internal,
+            cmd: cmd,
+            recipient: None,
+            player_state: None,
+            state: None,
+        }
+    }
+
+    fn recipient(mut self, r: &'a str) -> CommandSender {
+        self.recipient = Some(r);
+        self
+    }
+
+    fn player_state(mut self, s: &'a PlayerState) -> CommandSender {
+        self.player_state = Some(s);
+        self
+    }
+
+    fn state(mut self, s: protocol::spirc::State) -> CommandSender<'a> {
+        self.state = Some(s);
+        self
+    }
+
+    fn send(self) {
+        let mut pkt = protobuf_init!(protocol::spirc::Frame::new(), {
+            version: 1,
+            ident: self.spirc_internal.ident.clone(),
+            protocol_version: "2.0.0".to_owned(),
+            seq_nr: { self.spirc_internal.seq_nr += 1; self.spirc_internal.seq_nr  },
+            typ: self.cmd,
+            recipient: protobuf::RepeatedField::from_vec(
+                self.recipient.map(|r| vec![r.to_owned()] ).unwrap_or(vec![])
+                ),
+        });
+
+        if let Some(s) = self.player_state {
```

Comment:
Is there no way to refactor this block into one or is the borrow checker too much of a pain ?